includes/class-plugin-manager.php

```diff
@@ -0,0 +1,302 @@
+	 * @return string|bool Parsed slug on success. False on failure.
+	 */
+	public static function get_plugin_slug_from_url( $url ) {
+		if ( ! is_string( $url ) ) {
+			return false;
+		}
+
+		$url = wp_http_validate_url( $url );
+		if ( ! $url || ! stripos( $url, '.zip' ) ) {
+			return false;
+		}
+
+		$result = preg_match_all( '/\/([^\.\/*]+)/', $url, $matches );
+		if ( ! $result ) {
+			return false;
+		}
+
+		$group = end( $matches );
+		$slug  = end( $group );
+		return $slug;
+	}
+
+	/**
+	 * Installs a plugin.
+	 *
+	 * @param string $plugin Plugin slug or URL to plugin zip file.
+	 * @return bool True on success. False on failure.
+	 */
+	public static function install( $plugin ) {
+		if ( ! self::can_install_plugins() ) {
+			return new WP_Error( 'newspack_plugin_failed_install', __( 'Plugins cannot be installed.', 'newspack' ) );
+		}
+
+		if ( wp_http_validate_url( $plugin ) ) {
+			return self::install_from_url( $plugin );
+		} else {
+			return self::install_from_wporg( $plugin );
+		}
+	}
+
+	/**
+	 * Uninstall a plugin.
+	 *
+	 * @param string $plugin The plugin slug (e.g. 'newspack') or path to the plugin file. e.g. ('newspack/newspack.php').
+	 * @return bool True on success. False on failure.
+	 */
+	public static function uninstall( $plugin ) {
+		if ( ! self::can_install_plugins() ) {
+			return new WP_Error( 'newspack_plugin_failed_uninstall', __( 'Plugins cannot be uninstalled.', 'newspack' ) );
+		}
+
+		$installed_plugins = self::get_installed_plugins();
+		if ( ! in_array( $plugin, $installed_plugins ) && ! isset( $installed_plugins[ $plugin ] ) ) {
+			return new WP_Error( 'newspack_plugin_failed_uninstall', __( 'The plugin is not installed.', 'newspack' ) );
+		}
+
+		if ( isset( $installed_plugins[ $plugin ] ) ) {
+			$plugin_file = $installed_plugins[ $plugin ];
+		} else {
+			$plugin_file = $plugin;
+		}
+
+		// Deactivate plugin before uninstalling.
+		self::deactivate( $plugin_file );
+
+		$success = (bool) delete_plugins( [ $plugin_file ] );
+		if ( $success ) {
+			wp_clean_plugins_cache();
+			return true;
+		}
+		return new WP_Error( 'newspack_plugin_failed_uninstall', __( 'The plugin could not be uninstalled.', 'newspack' ) );
+	}
+
+	/**
+	 * Install a plugin from the WordPress.org plugin repo.
+	 *
+	 * @param string $plugin_slug The slug for the plugin.
+	 * @return bool True on success. False on failure.
+	 */
+	protected static function install_from_wporg( $plugin_slug ) {
+		// Quick check to make sure plugin directory doesn't already exist.
+		$plugin_directory = WP_PLUGIN_DIR . '/' . $plugin_slug;
+		if ( is_dir( $plugin_directory ) ) {
+			return new WP_Error( 'newspack_plugin_already_installed', __( 'The plugin directory already exists.', 'newspack' ) );
+		}
+
+		if ( ! function_exists( 'plugins_api' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/plugin-install.php';
+		}
+
+		$plugin_info = plugins_api(
+			'plugin_information',
+			[
+				'slug'   => $plugin_slug,
+				'fields' => [
```

Comment:
Are the fields necessary? I'm getting the same result with just:

```
$plugin_info = plugins_api(
	'plugin_information',
	[ 'slug' => $plugin_slug ]
);
```

Reply:
Those fields will limit the amount of information that the API call to WP.org returns. With the fields, the API will not transmit a bunch of extra info back. The end result is the same, since all the extra plugin info we don't need gets discarded, but it's faster with the fields.